cpp/serve/encoding.cc

```diff
@@ -0,0 +1,173 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file serve/encoding.cc
+ */
+#include "encoding.h"
+
+#include <tvm/runtime/logging.h>
+
+#include <array>
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+std::string CodepointToUtf8(TCodepoint codepoint) {
+  ICHECK(codepoint <= 0x10FFFF) << "Invalid codepoint: " << codepoint;
+  std::string utf8;
+  if (codepoint <= 0x7F) {
+    // 1-byte sequence
+    utf8 += static_cast<char>(codepoint);
+  } else if (codepoint <= 0x7FF) {
+    // 2-byte sequence
+    utf8 += static_cast<char>(0xC0 | ((codepoint >> 6) & 0x1F));
+    utf8 += static_cast<char>(0x80 | (codepoint & 0x3F));
+  } else if (codepoint <= 0xFFFF) {
+    // 3-byte sequence
+    utf8 += static_cast<char>(0xE0 | ((codepoint >> 12) & 0x0F));
+    utf8 += static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F));
+    utf8 += static_cast<char>(0x80 | (codepoint & 0x3F));
+  } else {
+    // 4-byte sequence
+    utf8 += static_cast<char>(0xF0 | ((codepoint >> 18) & 0x07));
+    utf8 += static_cast<char>(0x80 | ((codepoint >> 12) & 0x3F));
+    utf8 += static_cast<char>(0x80 | ((codepoint >> 6) & 0x3F));
+    utf8 += static_cast<char>(0x80 | (codepoint & 0x3F));
+  }
+  return utf8;
+}
+
+std::pair<TCodepoint, int> Utf8ToCodepoint(const char* utf8) {
```

Comment:
Just curious. Is the input `utf8` string guaranteed to be a valid UTF-8 string and we will never access illegal memory if otherwise?

Reply:
Good point. Added a guard to protect illegal memory access